### PR TITLE
Format list patterns

### PR DESCRIPTION
**Pull request by davidwengier**
Fixes https://github.com/dotnet/roslyn/issues/57244

Extract out a common base class from the `CurlyBraceCompletionService` so that parts can be shared with `BracketBraceCompeltionService` for list patterns, where we essentially want to format brackets as braces

Relates to test plan #51289

### CODE REVIEW

File: src/Features/CSharp/Portable/BraceCompletion/AbstractCurlyBraceOrBracketCompletionService.cs [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+using Microsoft.CodeAnalysis.Indentation;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.BraceCompletion
+{
+    internal abstract class AbstractCurlyBraceOrBracketCompletionService : AbstractBraceCompletionService
```
Comment (davidwengier) [inline src/Features/CSharp/Portable/BraceCompletion/AbstractCurlyBraceOrBracketCompletionService.cs:23]: This whole class is just a move from `CurlyBraceCompletionService` except where noted.

File: src/Features/CSharp/Portable/BraceCompletion/AbstractCurlyBraceOrBracketCompletionService.cs [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+            }
+
+            // Annotate the original closing brace so we can find it after formatting.
+            document = await GetDocumentWithAnnotatedClosingBraceAsync(document, closingPoint, cancellationToken).ConfigureAwait(false);
+
+            var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            var startPoint = openingPoint;
+            var endPoint = AdjustFormattingEndPoint(text, root, startPoint, closingPoint);
```
Comment (davidwengier) [inline src/Features/CSharp/Portable/BraceCompletion/AbstractCurlyBraceOrBracketCompletionService.cs:222]: The code in `CurlyBraceCompletionService.AdjustFormattingEndPoint` was previously inline at this point

File: src/Features/CSharp/Portable/BraceCompletion/AbstractCurlyBraceOrBracketCompletionService.cs [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+            var newClosingPoint = newRoot.GetAnnotatedTokens(s_closingBraceSyntaxAnnotation).Single().SpanStart + 1;
+
+            var textChanges = result.GetTextChanges(cancellationToken).ToImmutableArray();
+            return (textChanges, newClosingPoint);
+
+            async Task<Document> GetDocumentWithAnnotatedClosingBraceAsync(Document document, int closingBraceEndPoint, CancellationToken cancellationToken)
+            {
+                var originalRoot = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+                var closeBraceToken = originalRoot.FindToken(closingBraceEndPoint - 1);
+                Debug.Assert(IsValidClosingBraceToken(closeBraceToken));
```
Comment (davidwengier) [inline src/Features/CSharp/Portable/BraceCompletion/AbstractCurlyBraceOrBracketCompletionService.cs:262]: This used to directly check for `SyntaxKind.CloseBrace`

File: src/Features/CSharp/Portable/BraceCompletion/AbstractCurlyBraceOrBracketCompletionService.cs [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+
+                var newCloseBraceToken = closeBraceToken.WithAdditionalAnnotations(s_closingBraceSyntaxAnnotation);
+                var root = originalRoot.ReplaceToken(closeBraceToken, newCloseBraceToken);
+                return document.WithSyntaxRoot(root);
+            }
+        }
+
+        protected virtual int AdjustFormattingEndPoint(SourceText text, SyntaxNode root, int startPoint, int endPoint)
+            => endPoint;
```
Comment (jcouv) [inline src/Features/CSharp/Portable/BraceCompletion/AbstractCurlyBraceOrBracketCompletionService.cs:271]: Consider making this method abstract and add the override to `BracketBraceCompletionService`.

File: src/Features/CSharp/Portable/BraceCompletion/AbstractCurlyBraceOrBracketCompletionService.cs [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+    {
+        /// <summary>
+        /// Annotation used to find the closing brace location after formatting changes are applied.
+        /// The closing brace location is then used as the caret location.
+        /// </summary>
+        private static readonly SyntaxAnnotation s_closingBraceSyntaxAnnotation = new(nameof(s_closingBraceSyntaxAnnotation));
+
+        protected abstract ImmutableArray<AbstractFormattingRule> GetBraceFormattingIndentationRulesAfterReturn(DocumentOptionSet documentOptions);
+
+        public override async Task<BraceCompletionResult?> GetTextChangesAfterCompletionAsync(BraceCompletionContext braceCompletionContext, CancellationToken cancellationToken)
```
Comment (jcouv) [inline src/Features/CSharp/Portable/BraceCompletion/AbstractCurlyBraceOrBracketCompletionService.cs:33]: >override [](http://example.com/codeflow?start=15&length=8)

sealed?

File: src/Features/CSharp/Portable/BraceCompletion/AbstractCurlyBraceOrBracketCompletionService.cs [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+                if (!char.IsWhiteSpace(text[i]))
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        public override async Task<BraceCompletionResult?> GetTextChangeAfterReturnAsync(
```
Comment (jcouv) [inline src/Features/CSharp/Portable/BraceCompletion/AbstractCurlyBraceOrBracketCompletionService.cs:83]: >override [](http://example.com/codeflow?start=15&length=8)

sealed?

File: src/Features/CSharp/Portable/BraceCompletion/BracketBraceCompletionService.cs [hunk 7]
```diff
@@ -30,5 +37,41 @@ public override Task<bool> AllowOverTypeAsync(BraceCompletionContext context, Ca
         protected override bool IsValidOpeningBraceToken(SyntaxToken token) => token.IsKind(SyntaxKind.OpenBracketToken);
 
         protected override bool IsValidClosingBraceToken(SyntaxToken token) => token.IsKind(SyntaxKind.CloseBracketToken);
+
+        protected override ImmutableArray<AbstractFormattingRule> GetBraceFormattingIndentationRulesAfterReturn(DocumentOptionSet documentOptions)
+        {
+            return ImmutableArray.Create(BraceCompletionFormattingRule.Instance);
+        }
+
+        private sealed class BraceCompletionFormattingRule : BaseFormattingRule
```
Comment (jcouv) [inline src/Features/CSharp/Portable/BraceCompletion/BracketBraceCompletionService.cs:46]: >BraceCompletionFormattingRule [](http://example.com/codeflow?start=29&length=29)

Typo? Should this be called `BracketCompletionFormattingRule` or maybe just `FormattingRule`?

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/FormattingHelpers.cs [hunk 8]
```diff
@@ -48,6 +48,9 @@ public static string ContentBeforeLastNewLine(this IEnumerable<SyntaxTrivia> tri
         public static (SyntaxToken openBrace, SyntaxToken closeBrace) GetBracePair(this SyntaxNode? node)
             => node.GetBraces();
 
+        public static (SyntaxToken openBrace, SyntaxToken closeBrace) GetBracketPair(this SyntaxNode? node)
```
Comment (jcouv) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/FormattingHelpers.cs:51]: >openBrace [](http://example.com/codeflow?start=35&length=9)

Typo: `openBracket` and `closeBracket`?

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Rules/IndentBlockFormattingRule.cs [hunk 9]
```diff
@@ -246,6 +248,24 @@ private void AddBlockIndentationOperation(List<IndentBlockOperation> list, Synta
             AddIndentBlockOperation(list, bracePair.openBrace.GetNextToken(includeZeroWidth: true), bracePair.closeBrace.GetPreviousToken(includeZeroWidth: true));
         }
 
+        private static void AddBracketIndentationOperation(List<IndentBlockOperation> list, SyntaxNode node)
+        {
+            var bracketPair = node.GetBracketPair();
+
+            if (!bracketPair.IsValidBracePair())
```
Comment (jcouv) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Rules/IndentBlockFormattingRule.cs:255]: nit: Consider strengthening the logic in `IsValidBracePair` to check that we either have open-brace/close-brace or open-bracket/close-bracket. Maybe the name of the method should be tweaked? Maybe `IsValidBraceOrBracketPair`?

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 7). Didn't review the existing code that we merely moved

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Rules/IndentBlockFormattingRule.cs (hunk 9 above)
Comment (davidwengier) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Rules/IndentBlockFormattingRule.cs:253]: Good idea.. lets see what tests break :)

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 9)

**Pull request merged**